numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2191,40 +2191,33 @@ voidtype_item(PyVoidScalarObject *self, Py_ssize_t n)
     return voidtype_subscript(self, PyTuple_GetItem(flist, n));
 }
 
+static PyObject *
+gen_arrtype_subscript(PyObject *self, PyObject *key);
+
 /* get field by name or number */
 static PyObject *
 voidtype_subscript(PyVoidScalarObject *self, PyObject *ind)
 {
     npy_intp n;
     PyObject *ret, *args;
 
+    /* unstructured void scalars act like other scalars */
     if (!(PyDataType_HASFIELDS(self->descr))) {
-        PyErr_SetString(PyExc_IndexError,
-                "can't index void scalar without fields");
-        return NULL;
-    }
-
-#if defined(NPY_PY3K)
-    if (PyUString_Check(ind)) {
-#else
-    if (PyBytes_Check(ind) || PyUnicode_Check(ind)) {
-#endif
-        args = Py_BuildValue("(O)", ind);
-        ret = gentype_generic_method((PyObject *)self, args, NULL, "__getitem__");
-        Py_DECREF(args);
-        return ret;
+        return gen_arrtype_subscript((PyObject*)self, ind);
     }
 
-    /* try to convert it to a number */
+    /* structured scalar: try to convert it to a number */
     n = PyArray_PyIntAsIntp(ind);
-    if (error_converting(n)) {
-        goto fail;
+    if (!error_converting(n)) {
+        return voidtype_item(self, (Py_ssize_t)n);
     }
-    return voidtype_item(self, (Py_ssize_t)n);
+    PyErr_Clear();
 
-fail:
-    PyErr_SetString(PyExc_IndexError, "invalid index");
-    return NULL;
+    /* if it's not a number, try ndarray indexing */
+    args = Py_BuildValue("(O)", ind);
+    ret = gentype_generic_method((PyObject *)self, args, NULL, "__getitem__");
```

Comment:
Now that you mention it, I am going to change it to using  `__getitem__` even in in the unstructured case a few lines above too (will update). This better accounts for subclasses of `void` like `mvoid` which redefine `__getitem__`.

Using `gen_arrtype_subscript` actually breaks maskedarray printing, eg with your suggested change I get:
```python
>>> a = np.ma.array(np.zeros(2, dtype='i4,i4'), mask=[(0,0),(1,0)])
>>> a
masked_array(data = [(array(0, dtype=object), array(0, dtype=object))
 (array(--, dtype=object), array(0, dtype=object))],
             mask = [(False, False) ( True, False)],
       fill_value = (999999, 999999),
            dtype = [('f0', '<i4'), ('f1', '<i4')])
```

I will note that with the current PR (using `__getitem__` instead of  `gen_arrtype_subscript`), mvoid types don't work properly:
```python
>> a = np.ma.array(np.zeros(2, dtype='i4,i4'), mask=[(0,0),(1,0)])
>>> b = a[1]
>>> b[()]
masked
```
this is because `mvoid.__getitem__` isn't properly implemented to account for tuple indexing, since [this](https://github.com/numpy/numpy/blob/73c15f7b30658cbedc9c42e59b818b3d77dbc2c3/numpy/ma/core.py#L6034) line tests `m[indx]`, which returns a void `(1,0)` in this case. But honestly I don't want to try to fix maskedarray stuff in this PR if that's OK.